src/engine/sidechain/enginenetworkstream.cpp

```diff
@@ -0,0 +1,184 @@
+    }
+    int copyCount = math_min(writeAvailable, writeRequired);
+    if (copyCount > 0) {
+        (void)m_pOutputFifo->write(buffer, copyCount);
+        // we advance the frame only by the samples we have actually copied
+        // This means in case of buffer full (where we loose some frames)
+        // we do not get out of sync, and the syncing code tries to catch up the
+        // stream by writing silence, once the buffer is free.
+        m_streamFramesWritten += copyCount / m_numOutputChannels;
+    }
+    scheduleWorker();
+}
+
+void EngineNetworkStream::writeSilence(int frames) {
+    //qDebug() << "EngineNetworkStream::writeSilence()" << frames;
+    if (!m_pWorker->threadWaiting()) {
+        // no thread waiting, so we can advance the stream without
+        // buffering
+        m_streamFramesWritten += frames;
+        return;
+    }
+    int writeAvailable = m_pOutputFifo->writeAvailable();
+    int writeRequired = frames * m_numOutputChannels;
+    if (writeAvailable < writeRequired) {
+        qDebug() << "EngineNetworkStream::write() buffer full";
+    }
+    int clearCount = math_min(writeAvailable, writeRequired);
+    if (clearCount > 0) {
+        CSAMPLE* dataPtr1;
+        ring_buffer_size_t size1;
+        CSAMPLE* dataPtr2;
+        ring_buffer_size_t size2;
+        (void)m_pOutputFifo->aquireWriteRegions(clearCount,
+                &dataPtr1, &size1, &dataPtr2, &size2);
+        SampleUtil::clear(dataPtr1,size1);
+        if (size2 > 0) {
+            SampleUtil::clear(dataPtr2,size2);
+        }
+        m_pOutputFifo->releaseWriteRegions(clearCount);
+
+        // we advance the frame only by the samples we have actually cleared
+        m_streamFramesWritten += clearCount / m_numOutputChannels;
+    }
+    scheduleWorker();
+}
+
+void EngineNetworkStream::scheduleWorker() {
+    if (m_pOutputFifo->readAvailable()
+            >= m_numOutputChannels * kNetworkLatencyFrames) {
+        m_pWorker->outputAvailabe();
+    }
+}
+
+void EngineNetworkStream::read(CSAMPLE* buffer, int frames) {
+    int readAvailable = m_pOutputFifo->readAvailable();
+    int readRequired = frames * m_numInputChannels;
+    int copyCount = math_min(readAvailable, readRequired);
+    if (copyCount > 0) {
+        (void)m_pOutputFifo->read(buffer, copyCount);
+        buffer += copyCount;
+    }
+    if (readAvailable < readRequired) {
+        // Fill missing Samples with silence
+        int silenceCount = readRequired - readAvailable;
+        qDebug() << "EngineNetworkStream::write flushed" << readRequired
+                 << "samples";
+        SampleUtil::clear(buffer, silenceCount);
+    }
+}
+
+qint64 EngineNetworkStream::getStreamTimeFrames() {
+    return static_cast<double>(getStreamTimeUs()) * m_sampleRate / 1000000.0;
+}
+
+qint64 EngineNetworkStream::getStreamTimeUs() {
+    return getNetworkTimeUs() - m_streamStartTimeUs;
+}
+
+// static
+qint64 EngineNetworkStream::getNetworkTimeUs() {
```

Comment:
The CLOCK_MONOTONIC behaviour is exactly what we need. We need a clock that is synced by NTP frequency, not by an absolute value that might jump back. 

We have the following device chain: 
Mixxx -> Shoutcast Server -> Stream Client
All of them needs to be on the same sample rate, otherwise buffers somewhere in the chain will be slowly consumed or filled. 

The gettimeofday() in the current libschout implementation "works" because it is NTP synced as long no one set the wall clock.  

QueryPerformanceCounter, is based on the CPU crystal, which might float.
It looks like we need to stick on GetSystemTimeAsFileTime or timeGetTime and increase the precision by somehow by a QueryPerformanceCounter.
From Win8 we can use GetSystemTimePreciseAsFileTime 
 
All  others targets can use CLOCK_MONOTONIC